app/views.py

```diff
@@ -37,48 +38,45 @@ def home():
             return redirect(url_for('login'))
         login_user(user)
         return redirect(url_for('dashboard'))
+
     return render_template('index.html', loginForm=loginForm, codeForm=codeForm)
 
-@app.route('/progressreport/<int:knewbie_id>', methods=['GET','POST'])
+
+@app.route('/progressreport', methods=['GET', 'POST'])
 def get_report():
-    """Renders the class page for students/educators."""
+    """Renders the report page."""
     loginForm = LoginForm(prefix='login')
     codeForm = CodeForm(prefix='code')
+    if request.method == 'GET':
+        return redirect(url_for('progressreport'))
     if codeForm.validate_on_submit():
-        code = User.query.filter_by(knewbie_id=codeForm.code.data).first()
-        if code is None:
-            flash('Invalid code')
-            return redirect(url_for('home'))
-        return redirect(url_for('progressreport', knewbieID=codeForm.code.data))
-    return render_template('index.html', loginForm=loginForm, codeForm=codeForm)
+        return redirect(url_for('progressreport',knewbieID=codeForm.title.data))
+
+@app.route('/progressreport/')
+@app.route('/progressreport/<knewbieID>')
+def progressreport(knewbieID=None):
+    if knewbieID is None:
+        if not current_user.is_authenticated:
+            return redirect(url_for('login'))
+        elif current_user.urole == 'educator':
+            return redirect(url_for('dashboard'))
+        user = current_user
+    else:
+        user = User.query.filter_by(knewbie_id=knewbieID).first_or_404()
```

Comment:
Rather have a 404 Error instead of saying invalid code? 404 doesn't completely let them know the issue on hand

Reply:
Yep 404 here is fine, since the link access is via knewbieID.